src/class/proto_methods.rs

```diff
@@ -1,78 +1,135 @@
+}
+
+impl<T> PyDescrProtocolSlots<T> for &'_ PyClassProtocols<T> {
+    fn descr_protocol_slots(self) -> &'static [ffi::PyType_Slot] {
+        &[]
     }
 }
 
-#[doc(hidden)]
-#[cfg(feature = "macros")]
-pub trait PyProtoInventory: inventory::Collect {
-    fn new(methods: PyProtoMethodDef) -> Self;
-    fn get(&'static self) -> &'static PyProtoMethodDef;
+pub trait PyGCProtocolSlots<T> {
+    fn gc_protocol_slots(self) -> &'static [ffi::PyType_Slot];
 }
 
-#[doc(hidden)]
-#[cfg(feature = "macros")]
-pub trait HasProtoInventory {
-    type ProtoMethods: PyProtoInventory;
+impl<T> PyGCProtocolSlots<T> for &'_ PyClassProtocols<T> {
+    fn gc_protocol_slots(self) -> &'static [ffi::PyType_Slot] {
+        &[]
+    }
 }
 
-#[cfg(feature = "macros")]
-impl<T: HasProtoInventory> PyProtoMethods for T {
-    fn get_type_slots() -> Vec<ffi::PyType_Slot> {
-        inventory::iter::<T::ProtoMethods>
-            .into_iter()
-            .filter_map(|def| match def.get() {
-                PyProtoMethodDef::Slots(slots) => Some(slots),
-                PyProtoMethodDef::Buffer(_) => None,
-            })
-            .flatten()
-            .cloned()
-            .collect()
+pub trait PyIterProtocolSlots<T> {
+    fn iter_protocol_slots(self) -> &'static [ffi::PyType_Slot];
+}
+
+impl<T> PyIterProtocolSlots<T> for &'_ PyClassProtocols<T> {
+    fn iter_protocol_slots(self) -> &'static [ffi::PyType_Slot] {
+        &[]
     }
+}
+
+pub trait PyMappingProtocolSlots<T> {
+    fn mapping_protocol_slots(self) -> &'static [ffi::PyType_Slot];
+}
 
-    fn get_buffer() -> Option<PyBufferProcs> {
-        inventory::iter::<T::ProtoMethods>
-            .into_iter()
-            .find_map(|def| match def.get() {
-                PyProtoMethodDef::Slots(_) => None,
-                PyProtoMethodDef::Buffer(buf) => Some(buf.clone()),
-            })
+impl<T> PyMappingProtocolSlots<T> for &'_ PyClassProtocols<T> {
+    fn mapping_protocol_slots(self) -> &'static [ffi::PyType_Slot] {
+        &[]
+    }
+}
+
+pub trait PyNumberProtocolSlots<T> {
+    fn number_protocol_slots(self) -> &'static [ffi::PyType_Slot];
+}
+
+impl<T> PyNumberProtocolSlots<T> for &'_ PyClassProtocols<T> {
+    fn number_protocol_slots(self) -> &'static [ffi::PyType_Slot] {
+        &[]
+    }
+}
+
+pub trait PyAsyncProtocolSlots<T> {
+    fn async_protocol_slots(self) -> &'static [ffi::PyType_Slot];
+}
+
+impl<T> PyAsyncProtocolSlots<T> for &'_ PyClassProtocols<T> {
+    fn async_protocol_slots(self) -> &'static [ffi::PyType_Slot] {
+        &[]
+    }
+}
+
+pub trait PySequenceProtocolSlots<T> {
+    fn sequence_protocol_slots(self) -> &'static [ffi::PyType_Slot];
+}
+
+impl<T> PySequenceProtocolSlots<T> for &'_ PyClassProtocols<T> {
+    fn sequence_protocol_slots(self) -> &'static [ffi::PyType_Slot] {
+        &[]
+    }
+}
+
+// Buffer protocol has a slightly different implementation because the slots can't actually be used
+// to set the buffer protocol...
```

Comment:
Let's be more precise, like:
```
Initializing buffer protocol via `ffi::PyType_Slot` is supported in Python 3.9.1, so we have to manually set `PyBufferProcs`.
```

Reply:
👍 actually I went further and make `PyBufferProtocolSlots` work using slots for `#[cfg(Py_3_9)]`, so that maybe in a few years it's easy for us to clean up this hack 😂